File: test/refmd/crypto/suite_test.exs
defmodule RefMD.Crypto.SuiteTest do
  use ExUnit.Case, async: true

  alias RefMD.Crypto.Suite

  describe "suite admission" do
    test "validates current suite policy and pinned allowed-suite hash" do
      policy = Suite.current_suite_policy()
      assert policy["allowed_suite_ids_hash"] == "OcQ3VH6UrkTrIXcahgjG7weNblUpExxAM0rB5KqOVts"
      assert Suite.canonical_allowed_suite_ids_hash(policy) == policy["allowed_suite_ids_hash"]
      assert :ok = Suite.assert_protocol_version!(1)

      assert :ok =
               Suite.assert_known_suite_id!(
                 "refmd-v2-initial-ake-mlkem768-x25519-hkdfsha256-ed25519-mldsa65",
                 policy
               )

      assert :ok =
               Suite.assert_suite_rank_allowed!(
                 "refmd-v2-hybrid-signature-ed25519-mldsa65",
                 1000,
                 policy
               )

      assert :ok = Suite.assert_required_components!(policy)

      assert :ok =
               Suite.assert_pinned_suite_policy!(policy, %{
                 "suite_policy_version" => 1,
                 "min_suite_rank" => 1000,
                 "allowed_suite_ids_hash" => policy["allowed_suite_ids_hash"]
               })
    end

    test "rejects protocol and suite downgrades" do
      policy = Suite.current_suite_policy()

      assert_raise ArgumentError, fn -> Suite.assert_protocol_version!(0) end
      assert_raise ArgumentError, fn -> Suite.assert_protocol_version!(2) end

      assert_raise ArgumentError, fn ->
        Suite.assert_known_suite_id!("refmd-v1-static-dh", policy)
      end

      assert_raise ArgumentError, fn ->
        Suite.assert_suite_rank_allowed!("refmd-v2-hybrid-signature-ed25519-mldsa65", 999, policy)
      end

      assert_raise ArgumentError, fn ->
        Suite.assert_pinned_suite_policy!(
          %{policy | "allowed_suite_ids" => tl(policy["allowed_suite_ids"])},
          %{
            "suite_policy_version" => 1,
            "min_suite_rank" => 1000,
            "allowed_suite_ids_hash" => policy["allowed_suite_ids_hash"]
          }
        )
      end

      assert_raise ArgumentError, fn ->
        Suite.assert_known_suite_id!(
          "refmd-v2-unknown-extra",
          %{
            policy
            | "allowed_suite_ids" =>
                Enum.sort(policy["allowed_suite_ids"] ++ ["refmd-v2-unknown-extra"])
          }
        )
      end

      assert_raise ArgumentError, fn ->
        Suite.assert_required_components!(%{
          policy
          | "required_components" => ["ed25519", "mldsa65"]
        })
      end
    end
  end
end

File: assets/src/shared/lib/crypto/suite.test.ts
import { describe, expect, it } from "vitest";
import {
  assertKnownSuiteId,
  assertPinnedSuitePolicy,
  assertProtocolVersion,
  assertRequiredComponents,
  assertSuiteRankAllowed,
  canonicalAllowedSuiteIdsHash,
  currentSuitePolicy,
  SUITE_IDS,
} from "./suite";

describe("suite admission", () => {
  it("validates current suite policy and pinned allowed-suite hash", () => {
    const policy = currentSuitePolicy();
    expect(policy.allowed_suite_ids_hash).toBe("OcQ3VH6UrkTrIXcahgjG7weNblUpExxAM0rB5KqOVts");
    expect(canonicalAllowedSuiteIdsHash(policy)).toBe(policy.allowed_suite_ids_hash);
    expect(() => assertProtocolVersion(1)).not.toThrow();
    expect(() => assertKnownSuiteId(SUITE_IDS.INITIAL_AKE, policy)).not.toThrow();
    expect(() => assertSuiteRankAllowed(SUITE_IDS.HYBRID_SIGNATURE, 1000, policy)).not.toThrow();
    expect(() => assertRequiredComponents(policy)).not.toThrow();
    expect(() =>
      assertPinnedSuitePolicy(policy, {
        suite_policy_version: 1,
        min_suite_rank: 1000,
        allowed_suite_ids_hash: policy.allowed_suite_ids_hash!,
      }),
    ).not.toThrow();
  });

  it("rejects protocol and suite downgrades", () => {
    const policy = currentSuitePolicy();
    expect(() => assertProtocolVersion(0)).toThrow();
    expect(() => assertProtocolVersion(2)).toThrow();
    expect(() => assertKnownSuiteId("refmd-v1-static-dh", policy)).toThrow();
    expect(() => assertSuiteRankAllowed(SUITE_IDS.HYBRID_SIGNATURE, 999, policy)).toThrow();
    expect(() =>
      assertPinnedSuitePolicy(
        { ...policy, allowed_suite_ids: policy.allowed_suite_ids.slice(1) },
        {
          suite_policy_version: 1,
          min_suite_rank: 1000,
          allowed_suite_ids_hash: policy.allowed_suite_ids_hash!,
        },
      ),
    ).toThrow();
    expect(() =>
      assertKnownSuiteId("refmd-v2-unknown-extra", {
        ...policy,
        allowed_suite_ids: [...policy.allowed_suite_ids, "refmd-v2-unknown-extra"].sort(),
      }),
    ).toThrow();
    expect(() =>
      assertRequiredComponents({ ...policy, required_components: ["ed25519", "mldsa65"] }),
    ).toThrow();
  });
});
